kong/plugins/bot-detection/handler.lua

```diff
@@ -19,10 +44,79 @@ local function get_user_agent()
   return user_agent
 end
 
+local function examine_agent(user_agent, conf)
+  user_agent = strip(user_agent)
+
+  if conf.whitelist then
+    for _, rule in ipairs(conf.whitelist) do
+      if re_match(user_agent, rule) then
+        return MATCH_WHITELIST
+      end
+    end
+  end
+
+  if conf.blacklist then
+    for _, rule in ipairs(conf.blacklist) do
+      if re_match(user_agent, rule) then
+        return MATCH_BLACKLIST
+      end
+    end
+  end
+
+  for _, rule in ipairs(rules.bots) do
+    if re_match(user_agent, rule) then
+      return MATCH_BOT
+    end
+  end
+
+  return MATCH_EMPTY
+end
+
 function BotDetectionHandler:new()
   BotDetectionHandler.super.new(self, "bot-detection")
 end
 
+function BotDetectionHandler:init_worker()
+  local singletons    = require "kong.singletons"
+  local worker_events = singletons.worker_events
+  local dao_factory   = singletons.dao
+
+  -- load our existing plugins to create our cache spaces
+  local plugins, err = dao_factory.plugins:find_all({
+    name = "bot-detection",
+  })
+  if err then
+    ngx.log(ngx.ERR, "err in fetching plugins: ", err)
+  end
+
+  for i = 1, #plugins do
+    ua_cache[plugins[i].api_id] = {}
+  end
+
+  -- catch updates and tell each worker to adjust their cache
+  -- accordingly. for create/updates we create/purge the cache
+  -- by simply creating a new lua table; for deletions, as we no
+  -- longer need this cache data, we just remove it
+  worker_events.register(function(data)
+    if data.entity.name == "bot-detection" then
+      worker_events.post("bot-detection-invalidate", data.operation,
```

Comment:
It is correct indeed: the worker processing the Admin API update broadcast an event in the `invalidations` channel, every other node polls the event from a single worker, and that same worker, after purging the appropriate cache key, posts this `crud` event as a convenience for plugins. So in the end, each node gets notified, and because we are dealing with Lua-land caching here, each worker of each node must be notified as well, so this approach is correct!